library/Faktory/Worker.hs

```diff
@@ -59,69 +77,118 @@ instance ToJSON FailPayload where
+  => (Job arg -> IO ())
+  -> Worker ()
+processorLoop f = do
+  WorkerConfig{settings, workerSettings} <- ask
   let
     namespace = connectionInfoNamespace $ settingsConnection settings
     processAndAck job = do
-      mResult <- timeout (jobReserveForMicroseconds job) $ f job
+      mResult <- liftIO $ timeout (jobReserveForMicroseconds job) $ f job
       case mResult of
-        Nothing -> settingsLogError settings "Job reservation period expired."
-        Just () -> ackJob client job
+        Nothing -> liftIO $ settingsLogError settings "Job reservation period expired."
+        Just () -> ackJob job
 
-  emJob <- fetchJob client $ namespaceQueue namespace $ settingsQueue
+  emJob <- fetchJob $ namespaceQueue namespace $ settingsQueue
     workerSettings
 
   case emJob of
-    Left err -> settingsLogError settings $ "Invalid Job: " <> err
-    Right Nothing -> threadDelaySeconds $ settingsIdleDelay workerSettings
+    Left err -> liftIO $ settingsLogError settings $ "Invalid Job: " <> err
+    Right Nothing -> liftIO $ threadDelaySeconds $ settingsIdleDelay workerSettings
     Right (Just job) ->
       processAndAck job
         `catches` [ Handler $ \(ex :: WorkerHalt) -> throw ex
                   , Handler $ \(ex :: SomeException) ->
-                    failJob client job $ T.pack $ show ex
+                    failJob job $ T.pack $ show ex
                   ]
 
 -- | <https://github.com/contribsys/faktory/wiki/Worker-Lifecycle#heartbeat>
-heartBeat :: Client -> WorkerId -> IO ()
-heartBeat client workerId = do
-  threadDelaySeconds 25
-  command_ client "BEAT" [encode $ BeatPayload workerId]
+heartBeat :: Worker ()
+heartBeat = do
+  WorkerConfig{client, workerId} <- ask
+  liftIO $ threadDelaySeconds 25
+  liftIO $ command_ client "BEAT" [encode $ BeatPayload workerId]
 
 fetchJob
-  :: FromJSON args => Client -> Queue -> IO (Either String (Maybe (Job args)))
-fetchJob client queue = commandJSON client "FETCH" [queueArg queue]
-
-ackJob :: HasCallStack => Client -> Job args -> IO ()
-ackJob client job = commandOK client "ACK" [encode $ AckPayload $ jobJid job]
-
-failJob :: HasCallStack => Client -> Job args -> Text -> IO ()
-failJob client job message =
-  commandOK client "FAIL" [encode $ FailPayload message "" (jobJid job) []]
+  :: FromJSON args => Queue -> Worker (Either String (Maybe (Job args)))
+fetchJob queue = do
+  WorkerConfig{client} <- ask
+  liftIO $ commandJSON client "FETCH" [queueArg queue]
+
+ackJob :: HasCallStack => Job args -> Worker ()
+ackJob job = do
+  WorkerConfig{client} <- ask
+  liftIO $ commandOK client "ACK" [encode $ AckPayload $ jobJid job]
+
+failJob :: HasCallStack => Job args -> Text -> Worker ()
+failJob job message = do
+  WorkerConfig{client} <- ask
+  liftIO $ commandOK client "FAIL" [encode $ FailPayload message "" (jobJid job) []]
+
+untilM_ :: Monad m => m Bool -> m a -> m ()
+untilM_ predicate action =
+  predicate >>= \a -> unless a (action *> untilM_ predicate action)
```

Comment:
[Nit] You should probably swap this to do-notation since every other function in this file has that pattern.